Store coauthors by user_nicename, not by ID

goldenapples opened the pull request:
Addresses #9: because user ID's have the potential to overlap with post ID's (which are used as the ID for guest authors), this plugin will need to store the user_nicename instead.

Also, beefed up some of the tests in place and refactored out some of the copypasta from Co-Authors Plus.

goldenapples commented:
Apparently this conflicts with the branch I have open for the other pull request. Thought I was keeping them separate, but I will have to rebase this once the other one's merged in.

danielbachhuber commented on includes/author-roles-posts-relationships.php:
```diff
@@ -44,17 +46,18 @@ function set_contributor_on_post( $post_id, $author, $author_role = false ) {
+	}
+
+	if ( ! $author_role )
+		return false;
```
Bracing

danielbachhuber commented on includes/query.php:
```diff
@@ -92,32 +92,34 @@ function( $term ) { return 'cap-' . $term; },
+					// Since some authors may have also been added through postmeta, skip them here
+					// so as not to include duplicate authors in results.
+					if ( in_array( $coauthor_slug, wp_list_pluck( $coauthors, 'user_nicename' ) ) )
+						continue;
```
Bracing

danielbachhuber commented on includes/query.php:
```diff
@@ -92,32 +92,34 @@ function( $term ) { return 'cap-' . $term; },
+
+					// In case the user has been deleted while plugin was deactivated
+					if ( ! empty( $post_author ) )
+						$coauthors[] = $post_author;
```
Bracing

danielbachhuber commented:
Some small nits, but looking better. Doesn't seem to conflict though?

danielbachhuber commented on includes/admin-edit-ui.php:
```diff
@@ -104,11 +104,22 @@ function coauthors_meta_box( $post ) {
+	$coauthors = array_map(
+		function($author) {
+			if ( ! isset( $author->author_role ) )
+				$author->author_role = '';
```
Bracing

goldenapples commented:
Maybe github can merge it fine, but when I tried to merge one on top of the other locally, I get some errors:

```
Auto-merging tests/test-manage-contributors.php
CONFLICT (content): Merge conflict in tests/test-manage-contributors.php
Auto-merging includes/default-author-roles.php
Auto-merging includes/author-roles-posts-relationships.php
Auto-merging includes/admin-edit-ui.php
CONFLICT (content): Merge conflict in includes/admin-edit-ui.php
Automatic merge failed; fix conflicts and then commit the result.
```
Anyhow, it should be easy enough to clean up once one or the other is merged.  I'll fix these style issues.

goldenapples commented:
Fixed braces throughout. Want to look it over again?

danielbachhuber commented on includes/admin-edit-ui.php:
```diff
@@ -332,8 +349,9 @@ function search_coauthors( $search_term, $post_ID ) {
 		$coauthors = array_udiff( $coauthors, $existing,
 			function( $author, $existing ) {
-				if ( (int) $author->ID == (int) $existing->ID )
+				if ( (int) $author->user_nicename == (int) $existing->user_nicename ) {
```
Hm, why are you casting `user_nicename` to int?

danielbachhuber replied:
If not intentional, there probably should be a unit test that catches it.

goldenapples replied:
Ack. Yup. Will add a test that fails for that and fix it.

goldenapples commented:
OK, I added full test coverage for the `search_coauthors()` function and fixed the type coercion issue and two other bugs which turned up in unit testing.

danielbachhuber commented:
Looks good!

thanj commented:
Closes #9.

goldenapples commented:
Closes #9.
